cli/src/commands/package/upload.rs

```diff
@@ -0,0 +1,515 @@
+                                Err(_) if resume_on_error => {
+                                    statistics.add_failed_annotations(1);
+                                }
+                                Err(err) => error_sender
+                                    .send(anyhow::Error::msg(err))
+                                    .expect("Could not send error"),
+                            }
+                        }
+                    }
+
+                    Err(_) if resume_on_error => {
+                        statistics.add_failed_uploads(1);
+                    }
+
+                    Err(err) => error_sender.send(err).expect("Could not send error"),
+                }
+            })
+        }
+    });
+
+    drop(error_sender);
+
+    if let Ok(error) = error_receiver.try_recv() {
+        return Err(anyhow!("Failed to upload document {}", error));
+    };
+
+    Ok(())
+}
+
+pub fn run(args: &UploadPackageArgs, client: &Client, pool: &mut Pool) -> Result<()> {
+    let UploadPackageArgs {
+        file,
+        resume_on_error,
+        max_attachment_memory_mb,
+        dataset_creation_timeout,
+    } = args;
+
+    let mut package =
+        Package::new(file).context("Failed to read package, check path is correct.")?;
+
+    let statistics = Arc::new(Statistics::new());
+    let total_documents = package.get_document_count();
+
+    let _progress_bar = get_progress_bar(total_documents as u64, &statistics);
+
+    for dataset in package
+        .datasets()
+        .context("Could not get package datastes")?
+    {
+        let packaged_sources = get_ixp_source(&dataset, SourceProvider::Packaged(&mut package))
+            .context("Could not get ixp source from package")?;
+
+        let new_dataset = create_dataset(
+            dataset.name,
+            dataset.label_defs,
+            client,
+            *dataset_creation_timeout,
+        )
+        .context("Could not create dataset")?;
+
+        let new_sources = get_ixp_source(&new_dataset, SourceProvider::Remote(client))
+            .context("Could not get ixp source via api")?;
+
+        for (packaged, new) in [
+            (&packaged_sources.design_time, &new_sources.design_time),
+            (&packaged_sources.runtime, &new_sources.runtime),
+        ] {
+            unpack_ixp_source(
+                packaged,
+                new,
+                &new_dataset,
+                &mut package,
+                client,
+                pool,
+                *resume_on_error,
+                &statistics,
+                max_attachment_memory_mb,
+            )
+            .context("Failed to unpack ixp source")?;
+        }
+    }
+
+    log::info!("Package uploaded.");
+    Ok(())
+}
+
+#[derive(Debug)]
+pub struct Statistics {
+    document_reads: AtomicUsize,
+    document_uploads: AtomicUsize,
+    annotations: AtomicUsize,
+    failed_annotations: AtomicUsize,
+    failed_reads: AtomicUsize,
+    failed_uploads: AtomicUsize,
+}
```

Comment:
I feel like a lot of your stats could be shared with up/download with some generic naming like

```suggestion
#[derive(Debug)]
pub struct Statistics {
    document_package_count: AtomicUsize,
    document_api_count: AtomicUsize,
    annotations: AtomicUsize,
    failed_annotations: AtomicUsize,
    failed_package_count: AtomicUsize,
    failed_api_count: AtomicUsize,
}
```
as these modules are literally mirrors of each other.

But the names do get less readable, so not fussed.